plugin/pluginenv/environment.go

```diff
@@ -0,0 +1,123 @@
+// Package pluginenv provides high level functionality for discovering and launching plugins.
+package pluginenv
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/platform/plugin"
+)
+
+type APIProviderFunc func(*plugin.Manifest) (plugin.API, error)
+type SupervisorProviderFunc func(*plugin.BundleInfo) (plugin.Supervisor, error)
+
+// Environment represents an environment that plugins are discovered and launched in.
+type Environment struct {
+	searchPath         string
+	apiProvider        APIProviderFunc
+	supervisorProvider SupervisorProviderFunc
+	activePlugins      map[string]plugin.Supervisor
+}
+
+type Option func(*Environment)
+
+// Creates a new environment. At a minimum, the APIProvider and SearchPath options are required.
+func New(options ...Option) (*Environment, error) {
+	env := &Environment{
+		activePlugins: make(map[string]plugin.Supervisor),
+	}
+	for _, opt := range options {
+		opt(env)
+	}
+	if env.supervisorProvider == nil {
+		env.supervisorProvider = DefaultSupervisorProvider
+	}
+	if env.searchPath == "" {
+		return nil, fmt.Errorf("a search path must be provided")
+	} else if env.apiProvider == nil {
+		return nil, fmt.Errorf("an api provider must be provided")
+	}
+	return env, nil
+}
+
+// Returns a list of all plugins found within the environment.
+func (env *Environment) Plugins() ([]*plugin.BundleInfo, error) {
+	return ScanSearchPath(env.searchPath)
+}
+
+// Returns the ids of the currently active plugins.
+func (env *Environment) ActivePluginIds() (ids []string) {
```

Comment:
@jwilander I'm curious as to why we avoid them. I had never seen them prior to this code, and actually thought it was quite elegant